### PR TITLE
Salp

### PR DESCRIPTION
**Pull request by yrong**

### CODE REVIEW

File: pallets/salp/src/lib.rs [hunk 1]
```diff
@@ -196,28 +288,44 @@ pub mod pallet {
 		_,
 		Blake2_128Concat,
 		ParaId,
-		Option<FundInfo<T::AccountId, BalanceOf<T>, T::BlockNumber, LeasePeriodOf<T>>>,
-		ValueQuery
+		Option<FundInfo<AccountIdOf<T>, BalanceOf<T>, LeasePeriod>>,
```
Comment (yrong) [inline pallets/salp/src/lib.rs:319]: @AllenPocketGamer @ark930 seems LeasePeriod is not BlockNumber in crowdloan

File: pallets/salp/src/lib.rs [hunk 2]
```diff
@@ -262,16 +381,19 @@ pub mod pallet {
...
+			#[pallet::compact] last_slot: LeasePeriod,
 		) -> DispatchResultWithPostInfo {
 			let depositor = ensure_signed(origin)?;
 
 			ensure!(first_slot <= last_slot, Error::<T>::LastSlotBeforeFirstSlot);
-			let last_slot_limit = first_slot.checked_add(&3u32.into()).ok_or(Error::<T>::FirstSlotTooFarInFuture)?;
-			ensure!(last_slot <= last_slot_limit, Error::<T>::LastSlotTooFarInFuture);
-			ensure!(end > frame_system::Pallet::<T>::block_number(), Error::<T>::CannotEndInPast);
+			let last_slot_limit = first_slot
+				.checked_add(7u32.into())
```
Comment (yrong) [inline pallets/salp/src/lib.rs]: seems slot range is 8 in polkadot, need some review and confirm

File: pallets/salp/src/lib.rs [hunk 3]
```diff
@@ -306,112 +429,139 @@ pub mod pallet {
...
-		///
-		/// - `who`: The account whose contribution should be withdrawn.
+		/// Withdraw full balance of the parachain.
 		/// - `index`: The parachain to whose crowdloan the contribution was made.
-		// #[weight = T::WeightInfo::withdraw()]
 		#[pallet::weight(0)]
 		pub(super) fn withdraw(
 			origin: OriginFor<T>,
-			who: AccountId<T>,
 			#[pallet::compact] index: ParaId,
```
Comment (yrong) [inline pallets/salp/src/lib.rs:554]: the withdraw here is just to withdraw all raised from crowdloan account to parachain account with xcm transact

**Review by yrong — COMMENTED** [review]
add some more comments in salp

File: pallets/salp/src/lib.rs (hunk 1 above)
Comment (ark930) [reply pallets/salp/src/lib.rs:319]: LeasePeriod is defined as BlockNumber in pallet auctions

File: pallets/salp/src/lib.rs (hunk 2 above)
Comment (ark930) [reply pallets/salp/src/lib.rs]: total slot range is 8, include the first slot

**Review by ark930 — APPROVED** [review]

**Pull request merged**